File: admin/backend/api/v1/editor/routes/files.py
from __future__ import annotations

from flask import jsonify, request

from admin.backend.api.v1.editor import editor_bp, json_body, with_workspace


@editor_bp.get("/tree")
@with_workspace
def tree(ws):
    return jsonify(ws.tree(request.args.get("path", "")))


@editor_bp.get("/file")
@with_workspace
def read_file(ws):
    result = ws.read(request.args.get("path", ""))
    response = jsonify(result)
    response.headers["ETag"] = result["etag"]
    return response


@editor_bp.put("/file")
@with_workspace
def save_file(ws):
    status, result = ws.save(
        request.args.get("path", ""), json_body().get("content", ""), request.headers.get("If-Match", "*")
    )
    response = jsonify(result)
    response.status_code = status
    response.headers["ETag"] = result["etag"]
    return response


@editor_bp.post("/create")
@with_workspace
def create(ws):
    body = json_body()
    return jsonify(ws.create(body.get("path", ""), body.get("type", "file")))


@editor_bp.post("/rename")
@with_workspace
def rename(ws):
    body = json_body()
    return jsonify(ws.rename(body.get("from", ""), body.get("to", "")))


@editor_bp.delete("/delete")
@with_workspace
def delete(ws):
    ws.delete(request.args.get("path", ""))
    return jsonify({"ok": True})


@editor_bp.get("/files")
@with_workspace
def files(ws):
    return jsonify(ws.list_files())

File: admin/frontend/editor/src/api/files.js
import { request, body } from './client'

export const filesApi = {
  tree: (path = '') => request.get('tree', { searchParams: { path } }).then((r) => body(r)),
  list: () => request.get('files').then((r) => body(r, [])),
  create: (path, type) => request.post('create', { json: { path, type } }).then((r) => body(r)),
  rename: (from, to) => request.post('rename', { json: { from, to } }).then((r) => body(r)),
  del: (path) => request.delete('delete', { searchParams: { path } }).then((r) => body(r)),

  async read(path) {
    const res = await request.get('file', { searchParams: { path } })
    if (res.status === 415) throw new Error('binary file')
    return body(res)
  },

  async save(path, content, etag) {
    const res = await request.put('file', {
      searchParams: { path },
      headers: { 'If-Match': etag || '*' },
      json: { content },
    })
    if (res.status === 409) return { conflict: true, ...(await res.json()) }
    if (!res.ok) throw new Error((await res.text()) || 'Save failed')
    return { conflict: false, ...(await res.json()) }
  },
}
